Add regular polygons to visualize

arj0019 opened the pull request:
Solution for: #1019

Implements functionality for regular polygons by an alternate constructor function scope.

Example usage of the `polygon.regular` command:

```
#for k in range(3,21) {box(width: 36pt)[#polygon.regular(fill: purple.lighten(80%), stroke: purple, size: 30pt, vertices: k,)]}
```

![regular polygon](https://github.com/typst/typst/assets/92353079/896b9ab4-65a9-4dbf-9eac-4ce4a7fe25ee)

Topics to Discuss:

- The default size is `1em`; this does not match other shape implementations.
- ~~Some of the polygons generated are not centered in their debug box.~~
- Regular polygons with less than 3 vertices may be created.

arj0019 commented:
This PR is currently listed as a draft while other topics related to #1019 are discussed. It is not clear how the functionality of regular polygons should be implemented; the use of alternate constructors is being considered.

arj0019 commented:
I chose this implementation for a few reasons:

1. Alternate constructors are not currently standardized

2. The functionality of `square`, `rect`, `circle`, and `ellipse` is much closer to the desired functionality than `polygon`.

3. It appears that `circle` is really just a formatted call to `ellipse`. We could logically implement similar functionality for `square` and `triangle`.

PgBiel commented on library/src/visualize/shape.rs:
```diff
@@ -557,7 +671,15 @@ fn layout(
             let size = frame.size() + outset.sum_by_axis();
             let pos = Point::new(-outset.left, -outset.top);
-            let shape = ellipse(size, fill, stroke.left);
+            let shape: Shape;
+            match kind {
+              ShapeKind::Ngon(vertices) => {
+                shape = ngon(vertices, size, fill, stroke.left);
+              }
+              _ => {
+                shape = ellipse(size, fill, stroke.left);
+              }
+            }
```
Without getting into the discussion regarding the validity of this approach, here's a quick suggestion.

```suggestion
            let shape = match kind {
              ShapeKind::Ngon(vertices) => {
                ngon(vertices, size, fill, stroke.left)
              }
              _ => ellipse(size, fill, stroke.left)
            };
```

PgBiel commented:
By the way @arj0019 , I believe the conflict is due to https://github.com/typst/typst/commit/1fa56a317c8d8da76546e6b5e19b279d84380837, you might have to move your `define` of "ngon" to `visualize/mod.rs` now

arj0019 commented:
> By the way @arj0019 , I believe the conflict is due to [1fa56a3](https://github.com/typst/typst/commit/1fa56a317c8d8da76546e6b5e19b279d84380837), you might have to move your `define` of "ngon" to `visualize/mod.rs` now

What is the best way to do that? Can I pull the changes into my branch, and then make my change?

PgBiel commented:
> What is the best way to do that? Can I pull the changes into my branch, and then make my change?

I suggest making a separate branch with the latest changes (e.g. create a branch on your fork from the latest typst commit called `update-main`), then rebasing from it on your local computer with `git rebase update-main` (you will need to run `git pull` first to become aware of the new branch). Git will then give you the opportunity to resolve the conflicts manually by editing the relevant files (I suggest using a proper conflict resolver here, e.g. in some JetBrains IDE or `kdiff3` or some other program, to make it easier for you). You will then be able to `git rebase --continue` to apply your changes (in this case, moving the `define` to `visualize/mod.rs`, removing it from `lib.rs`). With this, you will have an altered branch with the conflicts resolved. You will need to `git push --force-with-lease` to apply the changes to your GitHub repository. **Ensure you didn't accidentally delete or break other changes, as this will overwrite all of your changes so far with the modified ones.** (Not that it would be that bad if you broke things since we have the history here in this PR anyway, this would be just to make everything easier.)

And that should be it (google "git rebase" for more detailed info). Hope this clears it up; feel free to share any further questions you might have in this process.

arj0019 commented:
Rebase is done, and I think everything still works...

laurmaedje commented:
I'm sorry for the lack of feedback on my part. I am still very unsure about `polygon.regular` and `ngon` and need more time to decide on the way forward.

arj0019 commented:
@laurmaedje I can start working on `polygon.regular` next week; the summer semester at school is currently wrapping up, and I will have a month off until next semester.

lnrs-industries commented:
@arj0019 can you give a status update? If you need help, I also need the polygon.regular and have some ideas for additional user-friendly features. Happy to take over the work if you don't have time to move forward :)

arj0019 commented:
> @arj0019 can you give a status update? If you need help, I also need the polygon.regular and have some ideas for additional user-friendly features. Happy to take over the work if you don't have time to move forward :)

I am pretty close to releasing the function scope variant of `polygon.regular`. IMO the `Shape` implementation is better than the `func` implementation, but this is what they wanted.

My only complication right now is with `Axes`, because `Axes::new(x, y)` requires `Rel<Length>` for the inputs which deforms the shape for a non-square `BoxElem`.

arj0019 commented:
The function scope variant of `polygon.regular` is ready for review.

@lnrs-industries, @PgBiel, @LaurenzV, @laurmaedje

PgBiel commented on crates/typst-library/src/visualize/polygon.rs:
```diff
@@ -19,13 +23,15 @@ use crate::prelude::*;
     /// Currently all polygons are filled according to the
     /// [non-zero winding rule](https://en.wikipedia.org/wiki/Nonzero-rule).
     pub fill: Option<Paint>,
+    pub test: Option<Paint>,
```
:raised_eyebrow:

arj0019 replied:
The test parameter has been removed.

PgBiel commented on crates/typst-library/src/visualize/polygon.rs:
```diff
@@ -91,3 +97,55 @@ impl Layout for PolygonElem {
     }
 }
+
+impl Reflect for PolygonElem {
+    fn describe() -> CastInfo {
+        CastInfo::Any
+    }
+
+    fn castable(_: &Value) -> bool {
+        false
+    }
+}
```
Any particular reason for this? (Perhaps the reasoning should be written as a comment above the `impl`)

arj0019 replied:
Reflect is required for the function scope; I will add a relevant comment.

```error[E0277]: the trait bound `polygon::PolygonElem: Reflect` is not satisfied
   --> crates\typst-library\src\visualize\polygon.rs:127:1
    |
127 | #[func]
    | ^^^^^^^ the trait `Reflect` is not implemented for `polygon::PolygonElem```

arj0019 replied:
Documentation for Reflect implementation justification has been added.

PgBiel commented on crates/typst-library/src/visualize/polygon.rs:
```diff
@@ -91,3 +97,55 @@ impl Layout for PolygonElem {
+    #[named] fill: Option<Paint>,
+    #[default(Smart::Auto)]
+    #[named]
+    stroke: Smart<Option<PartialStroke>>,
+    #[default(Length::from(Em::new(1.0)))]
+    #[named]
+    size: Length,
```
This is feeling a bit cluttered imo
```suggestion
    stroke: Smart<Option<PartialStroke>>,

    #[default(Length::from(Em::new(1.0)))]
    #[named]
    size: Length,

```

PgBiel commented on crates/typst-library/src/visualize/polygon.rs:
```diff
@@ -19,13 +23,15 @@ use crate::prelude::*;
 /// Display: Polygon
 /// Category: visualize
 #[element(Layout)]
+#[scope(scope.define("regular", regular_polygon_func()); scope)]
```
For consistency with other function scope declarations

```suggestion
#[scope(
    scope.define("regular", regular_polygon_func());
    scope
)]
```

PgBiel reviewed (commented):
initial review with some low-hanging fruits

arj0019 commented:
I have destroyed everything trying to update my files... I will try to fix it

arj0019 commented:
@PgBiel helped me fix my repo. Files have been modified to match suggestions.

arj0019 commented:
> I fixed some intricate issues in the implementation (polygon set rules would not have worked). However, a more "normal" problem still remains. The polygon isn't contained in its bounding box.
> 
> The following markup results in the image below:
> 
> ```
> #circle()
> #polygon.regular(vertices: 5)
> ```
> 
> ![regular polygon and circle](https://user-images.githubusercontent.com/17899797/261981462-63f64685-5278-43b8-b520-4e36111d49c6.png)
> 
> (I added `.debug()` to the frame returned by `PolygonElem::layout` which shows the blue bounding box and the red baseline.)
> 
> Both shapes have a size of 30pt, but the polygon is not contained in the bounding box because it's center is at (0, 0). Probably all points just need to be offset by half the size in both directions, but I haven't thought deeply about this.
> 
> Moreover, it would be great if you could add some tests to ensure that it works.

I will take a look at this and work on a solution.

arj0019 commented:
> Solution for: #1019
> 
> Implements functionality for regular polygons by an alternate constructor function scope.
> 
> Example usage of the `polygon.regular` command:
> 
> ```
> #for k in range(3,21) {box(width: 36pt)[#polygon.regular(fill: purple.lighten(80%), stroke: purple, size: 30pt, vertices: k,)]}
> ```
> 
> ![regular polygon debug](https://user-images.githubusercontent.com/92353079/263493117-fe1009a9-d688-4a5b-a113-e3e7a80aad02.png)
> 
> Topics to Discuss:
> 
> * The default size is `1em`; this does not match other shape implementations.
> * Some of the polygons generated are not centered in their debug box.
> * Regular polygons with less than 3 vertices may be created.

The current defaults for `shape`s appear random, and do not scale with a user defined size. The default size of `1em` scales with the user defined font size.

I am not sure why some of the regular polygons are not centered in the debug box. The math appears to be correct.

`polygon.regular` is an alternate constructor for `polygon`. If `polygon`s with less than 3 sides should not be allowed, then this should be implemented by `polygon` not `polygon.regular`.

arj0019 commented:
The center of the regular polygon has been moved horizontally so that the left most point is at `x = 0`.

PgBiel commented on tests/typ/visualize/polygon.typ:
```diff
@@ -8,6 +8,7 @@
 #polygon()
 #polygon((0em, 0pt))
 #polygon((0pt, 0pt), (10pt, 0pt))
+#polygon.regular(size: 0pt, vertices: 9)
```
more tests please (e.g. the regular polygon sheet you shared in the PR description)

arj0019 replied:
I added a test similar to the PR description.

PgBiel replied:
Thanks!

arj0019 commented on crates/typst-library/src/visualize/polygon.rs:
```diff
@@ -91,3 +97,67 @@ impl Layout for PolygonElem {
+    #[named]
+    stroke: Option<Smart<Option<PartialStroke>>>,
+
+    /// The size of the polygon.
```
The documentation has been updated.

```/// The diameter of the circumcircle of the regular polygon (https://en.wikipedia.org/wiki/Circumcircle).```

PgBiel replied:
Gotcha, thanks.

Design discussion alert: Would it perhaps not be better to specify the length of the side? It seems like a more tangible measurement when working with polygons.
Perhaps we could use some different opinions here.

arj0019 replied:
> It seems like a more tangible measurement when working with polygons.

I guess we think about it differently. IMO the circumcircle is the better way to define it. For a constant circumcircle diameter, varying the vertex count does not affect the size (outer bounds).

PgBiel replied:
Hmmm, well, you're not wrong in that sense. I wonder if it'd somehow be possible to specify both, but I'm not sure I'm capable of giving a proper opinion here. Cc @laurmaedje

laurmaedje replied:
I think size as circumcircle makes more sense than edge length. It's a bit unfortunate that the resulting bounding box isn't exactly size * size, but there's not much we can do about that. (I mean, we could adjust how polygon bboxes work and offset it slightly but it's also not really great.)

laurmaedje commented:
Thanks for your work on this!

arj0019 commented:
@PgBiel @laurmaedje Thanks for the help throughout this process. This was my first open source contribution. I enjoyed the challenge and learned a lot.
